poh/src/leader_bank_notifier.rs

```diff
@@ -0,0 +1,289 @@
+        drop(state);
+
+        self.condvar.notify_all();
+    }
+
+    /// Set the status to `StandBy` and notify any waiting threads if the status was changed.
+    /// Panics if the stored slot does not match the given slot.
+    pub fn set_completed(&self, slot: Slot) {
+        let mut state = self.state.lock().unwrap();
+        assert_eq!(state.slot, Some(slot));
+
+        // `set_completed` can be called multiple times for the same slot because it
+        // may be called from multiple threads.
+        if matches!(state.status, Status::StandBy) {
+            return;
+        }
+
+        state.status = Status::StandBy;
+        drop(state);
+
+        self.condvar.notify_all();
+    }
+
+    /// If the status is `InProgress`, immediately return a weak reference to the bank.
+    /// Otherwise, wait up to the `timeout` for the status to become `InProgress`.
+    /// If the timeout is reached, the weak reference is unupgradable.
+    pub fn get_or_wait_for_in_progress(&self, timeout: Duration) -> Weak<Bank> {
+        let state = self.state.lock().unwrap();
+        self.get_or_wait_for_in_progress_state(timeout, state)
+            .map(|state| state.bank.clone())
+            .unwrap_or_else(Weak::new)
+    }
+
+    /// Wait for next notification for a completed leader slot.
+    /// Returns `None` if the timeout is reached
+    pub fn wait_for_completed(&self, mut remaining_timeout: Duration) -> Option<Slot> {
+        let state = self.state.lock().unwrap();
+
+        // If currently `StandBy`, need to wait for `InProgress` to begin.
+        let now = Instant::now();
+        let state = self.get_or_wait_for_in_progress_state(remaining_timeout, state);
+
+        let Some(state) = state else { return None; };
+        remaining_timeout = remaining_timeout.saturating_sub(now.elapsed());
+
+        // Wait for `StandBy` to be set.
+        let (state, wait_timeout_result) = self
+            .condvar
+            .wait_timeout_while(state, remaining_timeout, |state| {
+                matches!(state.status, Status::InProgress)
+            })
+            .unwrap();
+
+        if !wait_timeout_result.timed_out() {
+            state.slot
+        } else {
+            None
+        }
+    }
+
+    // Helper function to get or wait for the `InProgress` status with a given `MutexGuard`.
+    fn get_or_wait_for_in_progress_state<'a>(
+        &self,
+        timeout: Duration,
+        state: MutexGuard<'a, SlotAndBankWithStatus>,
+    ) -> Option<MutexGuard<'a, SlotAndBankWithStatus>> {
+        // Immediately return if the status is `InProgress`
+        if matches!(state.status, Status::InProgress) {
+            return Some(state);
+        }
+
+        let (state, wait_timeout_result) = self
+            .condvar
+            .wait_timeout_while(state, timeout, |state| {
+                matches!(state.status, Status::StandBy)
+            })
+            .unwrap();
+
+        (!wait_timeout_result.timed_out()).then_some(state)
```

Comment:
nit: i think this 3 lines can be chained with `.filter_map()`

Reply:
I could do a filter_map on the `wait_timeout_while`, but I actually intended for a PoisonError on the mutex to panic with the unwrap. And the wait_timeout_result is unfortunately named, imo, as it is not really a `Result` that we can use the usual chaining methods on.

Reply:
oh, my bad. yeah, you're correct.